semgrep-core/src/analyzing/Constant_propagation.ml

```diff
@@ -496,23 +511,45 @@ let (terraform_stmt_to_vardefs : item -> (ident * expr) list) =
            | _ -> None)
   | _ -> []
 
+let (dockerfile_stmt_to_vardefs : item -> (ident * expr) list) =
+ fun st ->
+  let seen = Hashtbl.create 10 in
+  match st.s with
+  | ExprStmt
+      ({ e = Call ({ e = N (Id (("ENV", _), _)); _ }, (_, args, _)); _ }, _) ->
+      args
+      |> Common.map_filter (function
+           (* A shadowed variable cannot be constant, just ignore it *)
```

Comment:
I don't know too much about Dockerfiles, but can't they be constants? Or is it just that you're not sure about what's their exact value (but you know it's one of a few options) ?

Reply:
Yeah, I was going off of behavior like this in python, where once a variable gets shadowed, constant propagation gives up everywhere: https://semgrep.dev/s/R5Eq. But I see that it works inside a function: https://semgrep.dev/s/ALO2. I tried turning `ENV x=0 y=1` into a block of `ExprStmt (Assign ...)`, but this made it difficult to do matching correctly. I could spend more time trying to make it work if we think it's important.

Reply:
Actually, even when I translate `ENV` to assignments, I get the python toplevel behavior. Is there something special that would need to be done here?

Reply:
Ok, it's because of the dataflow phase of const-prop not running on dockerfile due to lack of function definitions. I think I can make this work.

Reply:
To be clear I was just trying to clarify if the comment was accurate. I think it's totally fine if you leave the code as it is but update the comment just to say that we don't handle those cases to keep it simple. In the dataflow const-prop indeed we handle them by using `Cst Cstr` which means "a constant of type string", which are useful in case the pattern is just `"..."`.

Reply:
Makes sense. I realized it was actually not too bad to do the translation using `DefStmt` without messing up matching.  That makes shadowing actually work. I'm going to close this and open a new PR.